File: web/lib/api/exercises.ts
import { apiFetch } from "./client";
import type { components } from "./schema";

export type MuscleGroup = components["schemas"]["MuscleGroupOut"];
export type Exercise = components["schemas"]["ExerciseOut"];
export type ExerciseCreate = components["schemas"]["ExerciseCreate"];
export type LoadType = components["schemas"]["LoadType"];
export type ProgressionMetric = components["schemas"]["ProgressionMetric"];
export type LastSession = components["schemas"]["LastSessionOut"];
export type SessionSummary = components["schemas"]["SessionOut"];
export type SetSummary = components["schemas"]["SetOut"];
export type HistoryPage = components["schemas"]["HistoryPageOut"];
export type ExercisePRs = components["schemas"]["ExercisePRsOut"];

export function listMuscleGroups(): Promise<MuscleGroup[]> {
  return apiFetch<MuscleGroup[]>("/muscle-groups");
}

export interface ListExercisesParams {
  q?: string;
  muscleGroup?: string;
}

export function listExercises(params: ListExercisesParams = {}): Promise<Exercise[]> {
  const search = new URLSearchParams();
  if (params.q) search.set("q", params.q);
  if (params.muscleGroup) search.set("muscle_group", params.muscleGroup);
  const qs = search.toString();
  return apiFetch<Exercise[]>(`/exercises${qs ? `?${qs}` : ""}`);
}

export function createExercise(data: ExerciseCreate): Promise<Exercise> {
  return apiFetch<Exercise>("/exercises", { method: "POST", body: data });
}

export function getExercise(exerciseId: string): Promise<Exercise> {
  return apiFetch<Exercise>(`/exercises/${exerciseId}`);
}

export function getLastSession(exerciseId: string): Promise<LastSession> {
  return apiFetch<LastSession>(`/exercises/${exerciseId}/last-session`);
}

export interface GetHistoryParams {
  limit?: number;
  cursor?: string;
}

export function getHistory(
  exerciseId: string,
  params: GetHistoryParams = {}
): Promise<HistoryPage> {
  const search = new URLSearchParams();
  if (params.limit) search.set("limit", String(params.limit));
  if (params.cursor) search.set("cursor", params.cursor);
  const qs = search.toString();
  return apiFetch<HistoryPage>(`/exercises/${exerciseId}/history${qs ? `?${qs}` : ""}`);
}

export function getPrs(exerciseId: string): Promise<ExercisePRs> {
  return apiFetch<ExercisePRs>(`/exercises/${exerciseId}/prs`);
}

export interface ExerciseLifetimeStats {
  sessionCount: number;
  totalVolumeGrams: number;
}

const LIFETIME_STATS_MAX_PAGES = 20;

/**
 * No backend aggregate exists for lifetime totals (Task 2.2 only shipped
 * last-session/history/prs), so this walks the cursor-paginated history and sums
 * it client-side. Bounded to avoid an unbounded loop for a pathological history size.
 */
export async function getExerciseLifetimeStats(
  exerciseId: string
): Promise<ExerciseLifetimeStats> {
  let sessionCount = 0;
  let totalVolumeGrams = 0;
  let cursor: string | undefined;

  for (let page = 0; page < LIFETIME_STATS_MAX_PAGES; page++) {
    const result = await getHistory(exerciseId, { limit: 100, cursor });
    sessionCount += result.sessions.length;
    totalVolumeGrams += result.sessions.reduce((sum, s) => sum + s.volume.grams, 0);
    if (!result.next_cursor) break;
    cursor = result.next_cursor;
  }

  return { sessionCount, totalVolumeGrams };
}

File: api/app/routers/exercises.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.dependencies import get_current_user, get_visible_exercise
from app.db.models import Exercise, MuscleGroup, User
from app.db.session import get_db
from app.schemas.analytics import ProgressionOut
from app.schemas.exercise import ExerciseCreate, ExerciseOut, ExerciseUpdate, MuscleGroupOut
from app.schemas.pr import ExercisePRsOut
from app.schemas.session import HistoryPageOut, LastSessionOut
from app.services import analytics_service, exercise_service
from app.services.analytics_service import Period

router = APIRouter(tags=["exercises"])


@router.get("/muscle-groups", response_model=list[MuscleGroupOut])
def list_muscle_groups(db: Session = Depends(get_db)) -> list[MuscleGroup]:
    return exercise_service.list_muscle_groups(db)


@router.get("/exercises", response_model=list[ExerciseOut])
def list_exercises(
    q: str | None = None,
    muscle_group: str | None = None,
    include_custom: bool = True,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> list[ExerciseOut]:
    exercises = exercise_service.list_exercises(db, current_user, q, muscle_group, include_custom)
    return [ExerciseOut.from_model(exercise) for exercise in exercises]


@router.post("/exercises", response_model=ExerciseOut)
def create_exercise(
    payload: ExerciseCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> ExerciseOut:
    exercise = exercise_service.create_exercise(db, current_user, payload)
    return ExerciseOut.from_model(exercise)


@router.get("/exercises/{exercise_id}", response_model=ExerciseOut)
def get_exercise(exercise: Exercise = Depends(get_visible_exercise)) -> ExerciseOut:
    return ExerciseOut.from_model(exercise)


@router.patch("/exercises/{exercise_id}", response_model=ExerciseOut)
def update_exercise(
    payload: ExerciseUpdate,
    exercise: Exercise = Depends(get_visible_exercise),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> ExerciseOut:
    updated = exercise_service.update_exercise(db, current_user, exercise, payload)
    return ExerciseOut.from_model(updated)


@router.get("/exercises/{exercise_id}/last-session", response_model=LastSessionOut)
def get_last_session(
    exercise: Exercise = Depends(get_visible_exercise),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> LastSessionOut:
    return exercise_service.get_last_session(db, current_user, exercise)


@router.get("/exercises/{exercise_id}/history", response_model=HistoryPageOut)
def get_history(
    limit: int = Query(default=20, ge=1, le=100),
    cursor: str | None = None,
    exercise: Exercise = Depends(get_visible_exercise),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> HistoryPageOut:
    return exercise_service.get_history(db, current_user, exercise, limit, cursor)


@router.get("/exercises/{exercise_id}/prs", response_model=ExercisePRsOut)
def get_prs(
    exercise: Exercise = Depends(get_visible_exercise),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> ExercisePRsOut:
    return exercise_service.get_prs(db, current_user, exercise)


@router.get("/exercises/{exercise_id}/progress", response_model=ProgressionOut)
def get_exercise_progress(
    period: Period = "90d",
    exercise: Exercise = Depends(get_visible_exercise),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> ProgressionOut:
    return analytics_service.get_exercise_progress(db, current_user, exercise, period)
